scientific_library/tvb/simulator/simulator.py

```diff
@@ -363,6 +422,42 @@ def _loop_monitor_output(self, step, state):
         if any(outputi is not None for outputi in output):
             return output
 
+    def bound_and_clamp(self, state):
+        # If there is a state boundary...
+        if self.integrator.state_variable_boundaries is not None:
+            # ...use the integrator's bound_state
+            self.integrator.bound_state(state)
+        # If there is a state clamping...
+        if self.integrator.clamped_state_variable_values is not None:
+            # ...use the integrator's clamp_state
+            self.integrator.clamp_state(state)
+
+    def _update_and_bound_history(self, history):
+        self.bound_and_clamp(history)
+        # If there are non-state variables, they need to be updated for history:
+        try:
+            update_initial_conditions = self.model.update_initial_conditions_non_state_variables
+        except:
+            try:
+                update_initial_conditions = self.model.update_non_state_variables
+            except:
+                return
+        # Assuming that node_coupling can have a maximum number of dimensions equal to the state variables,
+        # in the extreme case where all state variables are cvars as well, we set:
+        node_coupling = numpy.zeros((history.shape[0], 1, history.shape[2], self.model.number_of_modes))
+        for i_time in range(history.shape[1]):
+            update_initial_conditions(history[:, i_time], node_coupling[:, 0], 0.0)
+        self.bound_and_clamp(history)
+
+    def update_state(self, state, node_coupling, local_coupling=0.0, use_numba=True):
+        # If there are non-state variables, they need to be updated for the initial condition:
+        try:
+            self.model.update_non_state_variables
+        except:
```

Comment:
What exception would be raised here?  A bare `except` is poor style.

Reply:
No exception is to be raised here.
The try ... except... can be substituted by 
if there is an "model.update_non_state_variables" method, run it,
else, do nothing.